case_utils/case_validate/__init__.py

```diff
@@ -64,13 +65,175 @@ class NonExistentCDOConceptWarning(UserWarning):
+            (None, n_ontology_predicate, None)
+        ):
+            assert isinstance(ontology_triple[0], rdflib.URIRef)
+            assert isinstance(ontology_triple[2], rdflib.URIRef)
+            cdo_concepts.add(ontology_triple[0])
+            cdo_concepts.add(ontology_triple[2])
+    for ontology_triple in ontology_graph.triples((None, NS_RDF.type, NS_OWL.Ontology)):
+        if not isinstance(ontology_triple[0], rdflib.URIRef):
+            continue
+        cdo_concepts.add(ontology_triple[0])
+
+    # Also load historical ontology and version IRIs.
+    ontology_and_version_iris_data = importlib.resources.read_text(
+        case_utils.ontology, "ontology_and_version_iris.txt"
+    )
+    for line in ontology_and_version_iris_data.split("\n"):
+        cleaned_line = line.strip()
+        if cleaned_line == "":
+            continue
+        cdo_concepts.add(rdflib.URIRef(cleaned_line))
+
+    data_cdo_concepts: Set[rdflib.URIRef] = set()
+    for data_triple in data_graph.triples((None, None, None)):
+        for data_triple_member in data_triple:
+            if isinstance(data_triple_member, rdflib.URIRef):
+                if concept_is_cdo_concept(data_triple_member):
+                    data_cdo_concepts.add(data_triple_member)
+            elif isinstance(data_triple_member, rdflib.Literal):
+                if isinstance(data_triple_member.datatype, rdflib.URIRef):
+                    if concept_is_cdo_concept(data_triple_member.datatype):
+                        data_cdo_concepts.add(data_triple_member.datatype)
+
+    return data_cdo_concepts - cdo_concepts
+
+
+def validate(
+    input_file: str,
+    case_version: Optional[str] = None,
+    supplemental_graphs: Optional[List[str]] = None,
+    abort_on_first: bool = False,
+) -> ValidationResult:
+    """
+    Validate the given data graph against the given CASE ontology version and supplemental graphs.
+
+    :param input_file: The path to the file containing the data graph to validate.
+    :param case_version: The version of the CASE ontology to use.  If None, the most recent version will be used.
+    :param supplemental_graphs: The supplemental graphs to use.  If None, no supplemental graphs will be used.
+    :param abort_on_first: Whether to abort on the first validation error.
+    :return: The validation result object containing the defined properties.
+    """
+    # Convert the data graph string to a rdflib.Graph object.
+    data_graph = rdflib.Graph()
+    data_graph.parse(input_file)
+
+    # Get the ontology graph from the case_version and supplemental_graphs arguments
+    ontology_graph: Graph = get_ontology_graph(case_version, supplemental_graphs)
+
+    # Get the undefined CDO concepts
+    undefined_cdo_concepts = get_invalid_cdo_concepts(data_graph, ontology_graph)
+
+    # Validate data graph against ontology graph.
+    validate_result: Tuple[
+        bool, Union[Exception, bytes, str, rdflib.Graph], str
```

Comment:
It might be worth checking upstream with the `pyshacl` folks on what this signature should be.  An `Exception` can be returned? not raised??

Reply:
In fairness to this PR, this `Tuple` is the current signature (including the `Exception`).

Reply:
I followed the existing logic to not change too much, though we can change it in a future PR

Reply:
Type review issues are being addressed upstream, including their [PR 192](https://github.com/RDFLib/pySHACL/pull/192).  Let's not consider the type signature on `pyshacl.validate.validate` a blocking issue.  (IIRC, it might be blocked by PR 192.)